File: commandPattern/command.cpp
#include <iostream>
#include "./command.h"
#include "./receiver.h"

void HelloCommand::Execute()
{
	std::cout << "Hello Command" << std::endl;
}

WorldCommand::WorldCommand(Receiver *receiver)
{
	_receiver = receiver;
}

void WorldCommand::Execute()
{
	std::cout << "World Command" << std::endl;
	_receiver->ReceiverOperation();
}

void MarcoCommand::AddCommand(Command *command)
{
	_commands->push_front(command);
}

void MarcoCommand::RemoveCommand(Command *command)
{
	_commands->remove(command);
}

void MarcoCommand::Execute()
{
	list<Command *>::iterator it;

	for (it = _commands->begin(); it != _commands->end(); ++it)
	{
		Command *currentCommand = *it;
		currentCommand->Execute();
	}
}

File: commandPattern/csharp/Command.cs
abstract class Command
{
	public abstract void Execute();
}

class HelloCommand : Command
{
	public override void Execute()
	{
		System.Console.WriteLine("Hello Command");
	}
}

class WorldCommand : Command
{
	private Receiver _receiver;

	public WorldCommand(Receiver receiver)
	{
		_receiver = receiver;
	}

	public override void Execute()
	{
		System.Console.WriteLine("World Command");
		_receiver.ReceiverOperation();
	}
}

class MarcoCommand : Command
{
	private List<Command> commands = new List<Command>();

	public void AddCommand(Command command)
	{
		commands.Add(command);
	}


	public void RemoveCommand(Command command)
	{
		commands.Remove(command);
	}

	public override void Execute()
	{
		commands.ForEach(delegate (Command command)
		{
			command.Execute();
		});
	}
}